mbedtls/src/ssl/async_io.rs

```diff
@@ -0,0 +1,184 @@
+/* Copyright (c) Fortanix, Inc.
+ *
+ * Licensed under the GNU General Public License, version 2 <LICENSE-GPL or
+ * https://www.gnu.org/licenses/gpl-2.0.html> or the Apache License, Version
+ * 2.0 <LICENSE-APACHE or http://www.apache.org/licenses/LICENSE-2.0>, at your
+ * option. This file may not be copied, modified, or distributed except
+ * according to those terms. */
+
+#![cfg(all(feature = "std", feature = "async"))]
+
+use crate::{
+    error::{Error, Result},
+    ssl::{
+        context::Context,
+        io::{IoCallback, IoCallbackUnsafe},
+    },
+};
+use std::{
+    future::Future,
+    io::{Error as IoError, ErrorKind as IoErrorKind, Result as IoResult},
+    pin::Pin,
+    task::{Context as TaskContext, Poll},
+};
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
+
+/// Marker type for an IO implementation that implements both
+/// `tokio::io::AsyncRead` and `tokio::io::AsyncWrite`.
+pub enum AsyncStream {}
+
+// TODO: Add enum `AnyAsyncIo` as marker type for an IO implementation that
+// doesn't implement `tokio::io::AsyncRead` and `tokio::io::AsyncWrite`.
+
+// TODO: Add `AsyncIo` trait for async IO that that doesn't implement
+// `tokio::io::AsyncRead` and `tokio::io::AsyncWrite`. For example:
+//     pub trait AsyncIo {
+//        async fn recv(&mut self, buf: &mut [u8]) -> Result<usize>;
+//         async fn send(&mut self, buf: &[u8]) -> Result<usize>;
+//     }
+// Could implement by using `async-trait` crate or
+// #![feature(async_fn_in_trait)] or Associated Types
+
+impl<'a, 'b, 'c, IO: AsyncRead + AsyncWrite + std::marker::Unpin + 'static> IoCallback<AsyncStream>
+    for (&'a mut TaskContext<'b>, &'c mut IO)
+{
+    fn recv(&mut self, buf: &mut [u8]) -> Result<usize> {
+        let mut buf = ReadBuf::new(buf);
+        let io = Pin::new(&mut self.1);
+        match io.poll_read(self.0, &mut buf) {
+            Poll::Ready(Ok(())) => Ok(buf.filled().len()),
+            Poll::Ready(Err(_)) => Err(Error::NetRecvFailed),
+            Poll::Pending => Err(Error::SslWantRead),
+        }
+    }
+
+    fn send(&mut self, buf: &[u8]) -> Result<usize> {
+        let io = Pin::new(&mut self.1);
+        match io.poll_write(self.0, buf) {
+            // Need to handle 0 return especially because:
+            // A return value of 0 typically means that the underlying object is no longer able to accept bytes and will likely
+            // not be able to in the future as well, or that the buffer provided is empty.
+            //
+            // This is important because:
+            // c-mbedtls assumes: the callback must return the number of bytes sent if any, or a non-zero error code.
+            //
+            // Ref: https://docs.rs/tokio/latest/tokio/io/trait.AsyncWrite.html#tymethod.poll_write
+            // Ref: https://github.com/fortanix/rust-mbedtls/blob/master/mbedtls-sys/vendor/include/mbedtls/ssl.h#L572-L591
+            Poll::Ready(Err(_)) | Poll::Ready(Ok(0)) => Err(Error::NetSendFailed),
```

Comment:
Why is 0 handled as a special case in async only?

Reply:
I see, originally when I read C code of `mbedtls_ssl_flush_output`, I found it will return success code when callback return 0. So I think it might cause problem. But it seems the code there is correct.
This `0` need to be handled by out most caller

Reply:
reverted to original way